kitsune/wiki/models.py

```diff
@@ -263,39 +279,53 @@ def save(self, *args, **kwargs):
         # which would cause a rollback, which would negate any category changes
         # we make here, so don't worry:
         self._clean_category()
+        self._clean_template_status()
 
         slug_changed = hasattr(self, 'old_slug')
         title_changed = hasattr(self, 'old_title')
 
-        # If the slug changed, we clear out the share link so it gets regenerated.
-        self.share_link = ''
+        templateness_changed = (
+            title_changed and
+            # If the title was a template before, and is not now, or vice-versa
+            (self.title.startswith(TEMPLATE_TITLE_PREFIX) !=
+             self.old_title.startswith(TEMPLATE_TITLE_PREFIX)))
+
+        if slug_changed:
+            # Clear out the share link so it gets regenerated.
+            self.share_link = ''
 
         super(Document, self).save(*args, **kwargs)
 
         # Make redirects if there's an approved revision and title or slug
         # changed. Allowing redirects for unapproved docs would (1) be of
         # limited use and (2) require making Revision.creator nullable.
-        if self.current_revision and (slug_changed or title_changed):
+        #
+        # If the "templateness" changed, don't try and make a redirect.
+        # The category and the title-prefix have to stay consistent, and
+        # redirects are in the same category as their targets. There is not
+        # a clear way to construct the redirect without invalidating one of
+        # those conditions.
+        if self.current_revision and (slug_changed or title_changed) and not templateness_changed:
```

Comment:
I don't know this stuff well enough to know whether there are issues with
that plan, but it sounds fine as far as I can tell.

Is there anyone you can talk to whip might have a better idea? If not, I
vote go with it and see whether the tests show any issues.
On Jul 23, 2015 3:55 PM, "Mike Cooper" <notifications@github.com> wrote:

> In kitsune/wiki/models.py
> <https://github.com/mozilla/kitsune/pull/2547#discussion_r35364001>:
>
> >
> >          super(Document, self).save(*args, **kwargs)
> >
> >          # Make redirects if there's an approved revision and title or slug
> >          # changed. Allowing redirects for unapproved docs would (1) be of
> >          # limited use and (2) require making Revision.creator nullable.
> > -        if self.current_revision and (slug_changed or title_changed):
> > +        #
> > +        # If the "templateness" changed, don't try and make a redirect.
> > +        # The category and the title-prefix have to stay consistent, and
> > +        # redirects are in the same category as their targets. There is not
> > +        # a clear way to construct the redirect without invalidating one of
> > +        # those conditions.
> > +        if self.current_revision and (slug_changed or title_changed) and not templateness_changed:
>
> We had redirects for templates before. I wanted to keep that behavior (to
> minimize change) where it made sense. It doesn't make sense to make the
> redirects when documents switch to templates or vice versa (for the reasons
> explained in the comment).
>
> Thinking about this more though, I don't think having a template redirect
> would actually do anything useful. The template code doesn't follow
> redirects, so at best you would get some strange content in the article
> that includes the template. I think I'll change this to simply never add
> redirects for templates, which will simplify some things. Does that sound
> right?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mozilla/kitsune/pull/2547/files#r35364001>.
>